File: vss-mobile/android/src/main/java/com/vss/simulator/SimulatorBridge.java
package com.vss.simulator;

import android.content.res.AssetManager;
import android.util.Log;
import android.view.Surface;

/**
 * Bridge for accessing the simulator's native library.
 */
public class SimulatorBridge {
    private static final String LOG_TAG = "SimulatorBridge";
    private static boolean LIBRARY_LOADED = false;

    static {
        try {
            Log.d(LOG_TAG, "Loading native library...");
            System.loadLibrary("vss");
            LIBRARY_LOADED = true;
            Log.i(LOG_TAG, "Loading native library: successful");
        } catch (java.lang.UnsatisfiedLinkError e) {
            Log.e(LOG_TAG, "Loading native library: failed", e);
        }
    }

    private static native void nativeCreate(Surface surface, AssetManager assetManager);

    private static native void nativeDestroy();

    private static native void nativeResize(int width, int height);

    private static native void nativeDraw();

    private static native void nativePostFrame(int width, int height, byte[] y, byte[] u, byte[] v);

    private static native void nativePostSettings(String jsonString);

    public static boolean hasLoadedLibrary() {
        return LIBRARY_LOADED;
    }

    public static void create(Surface surface, AssetManager assetManager) {
        assert LIBRARY_LOADED : "Native library not loaded";
        Log.d(LOG_TAG, "Creating simulator");
        nativeCreate(surface, assetManager);
    }

    public static void destroy() {
        assert LIBRARY_LOADED : "Native library not loaded";
        Log.d(LOG_TAG, "Destroying simulator");
        nativeDestroy();
    }

    public static void resize(int width, int height) {
        assert LIBRARY_LOADED : "Native library not loaded";
        Log.v(LOG_TAG, "Resizing simulation to " + width + "x" + height);
        nativeResize(width, height);
    }

    public static void draw() {
        assert LIBRARY_LOADED : "Native library not loaded";
        Log.v(LOG_TAG, "Drawing simulation");
        nativeDraw();
    }

    public static void postFrame(int width, int height, byte[] y, byte[] u, byte[] v) {
        assert LIBRARY_LOADED : "Native library not loaded";
        Log.v(LOG_TAG, "Posting input frame");
        nativePostFrame(width, height, y, u, v);
    }

    public static void postSettings(String jsonString) {
        assert LIBRARY_LOADED : "Native library not loaded";
        Log.v(LOG_TAG, "Posting simulator settings: " + jsonString);
        nativePostSettings(jsonString);
    }
}

File: vss-android/app/src/main/kotlin/com/vss/simulator/SimulatorBridge.kt
package com.vss.simulator

import android.content.res.AssetManager
import android.hardware.HardwareBuffer
import android.util.Log
import android.view.Surface
import java.nio.ByteBuffer

/** Bridge for accessing the simulator's native library. */
object SimulatorBridge {
    private const val LOG_TAG = "SimulatorBridge"

    private var libraryLoaded = false

    init {
        try {
            Log.d(LOG_TAG, "Loading native library...")
            System.loadLibrary("vss_android")
            libraryLoaded = true
            Log.i(LOG_TAG, "Loading native library: successful")
        } catch (error: UnsatisfiedLinkError) {
            Log.e(LOG_TAG, "Loading native library: failed", error)
        }
    }

    @JvmStatic
    private external fun nativeCreate(surface: Surface, assetManager: AssetManager)

    @JvmStatic
    private external fun nativeDestroy()

    @JvmStatic
    private external fun nativeResize(width: Int, height: Int)

    @JvmStatic
    private external fun nativeDraw()

    @JvmStatic
    private external fun nativePostHardwareBuffer(
        width: Int,
        height: Int,
        dataSpace: Int,
        rotationDegrees: Int,
        hardwareBuffer: HardwareBuffer,
    )

    @JvmStatic
    private external fun nativePostRgba(width: Int, height: Int, pixels: ByteBuffer)

    @JvmStatic
    private external fun nativePostSettings(jsonString: String)

    @JvmStatic
    private external fun nativeQuerySettings(): String

    @JvmStatic
    private external fun nativeCatalog(locale: String): String

    @JvmStatic
    private external fun nativeComposeSettings(
        locale: String,
        activeProfiles: String,
        manualOverrides: String,
    ): String

    fun hasLoadedLibrary(): Boolean = libraryLoaded

    fun create(surface: Surface, assetManager: AssetManager) {
        assert(libraryLoaded) { "Native library not loaded" }
        Log.v(LOG_TAG, "Creating simulator")
        nativeCreate(surface, assetManager)
    }

    fun destroy() {
        assert(libraryLoaded) { "Native library not loaded" }
        Log.v(LOG_TAG, "Destroying simulator")
        nativeDestroy()
    }

    fun resize(width: Int, height: Int) {
        assert(libraryLoaded) { "Native library not loaded" }
        Log.v(LOG_TAG, "Resizing simulation to ${width}x$height")
        nativeResize(width, height)
    }

    fun draw() {
        assert(libraryLoaded) { "Native library not loaded" }
        nativeDraw()
    }

    fun postHardwareBuffer(
        width: Int,
        height: Int,
        dataSpace: Int,
        rotationDegrees: Int,
        hardwareBuffer: HardwareBuffer,
    ) {
        assert(libraryLoaded) { "Native library not loaded" }
        nativePostHardwareBuffer(width, height, dataSpace, rotationDegrees, hardwareBuffer)
    }

    fun postRgba(width: Int, height: Int, pixels: ByteBuffer) {
        assert(libraryLoaded) { "Native library not loaded" }
        nativePostRgba(width, height, pixels)
    }

    fun postSettings(jsonString: String) {
        assert(libraryLoaded) { "Native library not loaded" }
        Log.v(LOG_TAG, "Posting simulator settings: $jsonString")
        nativePostSettings(jsonString)
    }

    fun querySettings(): String {
        assert(libraryLoaded) { "Native library not loaded" }
        Log.v(LOG_TAG, "Querying simulator settings")
        return nativeQuerySettings()
    }

    fun catalog(locale: String): String {
        assert(libraryLoaded) { "Native library not loaded" }
        return nativeCatalog(locale)
    }

    fun composeSettings(
        locale: String,
        activeProfiles: String,
        manualOverrides: String,
    ): String {
        assert(libraryLoaded) { "Native library not loaded" }
        return nativeComposeSettings(locale, activeProfiles, manualOverrides)
    }
}
